packages/analytics/shippers/elastic_v3/server/src/server_shipper.ts

```diff
@@ -0,0 +1,286 @@
+
+  private readonly reportTelemetryCounters = createTelemetryCounterHelper(
+    this.telemetryCounter$,
+    ElasticV3ServerShipper.shipperName
+  );
+
+  private readonly internalQueue: Event[] = [];
+  private readonly shutdown$ = new Subject<void>();
+
+  private lastBatchSent = Date.now();
+
+  private clusterUuid: string = 'UNKNOWN';
+  private licenseId?: string;
+  private isOptedIn?: boolean;
+
+  /**
+   * Specifies when it went offline:
+   * - `undefined` means it doesn't know yet whether it is online or offline
+   * - `null` means it's online
+   * - `number` means it's offline since that time
+   * @private
+   */
+  private firstTimeOffline?: number | null;
+
+  constructor(
+    private readonly options: ElasticV3ShipperOptions,
+    private readonly initContext: AnalyticsClientInitContext
+  ) {
+    this.setInternalSubscriber();
+    this.checkConnectivity();
+  }
+
+  public extendContext(newContext: EventContext) {
+    if (newContext.cluster_uuid) {
+      this.clusterUuid = newContext.cluster_uuid;
+    }
+    if (newContext.license_id) {
+      this.licenseId = newContext.license_id;
+    }
+  }
+
+  public optIn(isOptedIn: boolean) {
+    this.isOptedIn = isOptedIn;
+
+    if (isOptedIn === false) {
+      this.internalQueue.length = 0;
+    }
+  }
+
+  public reportEvents(events: Event[]) {
+    if (
+      this.isOptedIn === false ||
+      (this.firstTimeOffline && Date.now() - this.firstTimeOffline > 24 * HOUR)
+    ) {
+      return;
+    }
+
+    const freeSpace = 1000 - this.internalQueue.length;
+
+    // As per design, we only want store up-to 1000 events at a time. Drop anything that goes beyond that limit
+    if (freeSpace < events.length) {
+      const droppedEvents = events.slice(-(events.length - freeSpace));
+      events.length = freeSpace;
+      this.reportTelemetryCounters(droppedEvents, {
+        type: TelemetryCounterType.dropped,
+        code: 'queue_full',
+      });
+    }
+
+    this.internalQueue.push(...events);
+  }
+
+  public shutdown() {
+    this.shutdown$.complete();
+  }
+
+  /**
+   * Checks the server has connectivity to the remote endpoint.
+   * The frequency of the connectivity tests will back off, starting with 1 minute, and multiplying by 2
+   * until it reaches 1 hour. Then, it’ll keep the 1h frequency until it reaches 24h without connectivity.
+   * At that point, it clears the queue and stops accepting events in the queue.
+   * The connectivity checks will continue to happen every 1 hour just in case it regains it at any point.
+   * @private
+   */
+  private checkConnectivity() {
+    let backoff = 1 * MINUTE;
+    timer(0, 1 * MINUTE)
+      .pipe(
+        takeWhile(() => this.shutdown$.isStopped === false),
+        filter(() => this.isOptedIn === true && this.firstTimeOffline !== null),
+        concatMap(async () => {
+          const { ok } = await fetch(buildUrl(this.initContext.sendTo, this.options.channelName), {
```

Comment:
NIT: the `buildUrl` call can be extracted out to the top level of the function

Reply:
I went ahead and tackled https://github.com/elastic/kibana/issues/129014 as well with the related change.